sockeye/layers.py

```diff
@@ -68,8 +69,205 @@ def normalize(self, inputs: mx.sym.Symbol, eps: float = 0.00001) -> mx.sym.Symbo
+        self.b_h2o = mx.sym.Variable("%sh2o_bias" % prefix)
+        self.use_loop = False  # use naive loop
+
+    def on(self, inputs: mx.sym.Symbol, inputs_length: mx.sym.Symbol, length: int) -> mx.sym.Symbol:
+        """
+        Returns a symbol of shape (batch, length, output_depth).
+
+        :param inputs: Symbol of shape (batch, length, input_depth).
+        :param inputs_length: Symbol of shape (batch, 1).
+        :param length: Size of time dimension.
+        :return: Symbol of shape (batch, length, output_depth).
+        """
+        # lets do self attention first (code is constrained to q length == k length)
+        # inputs: (batch, length, num_hidden)
+
+        # Combine batch and time dimension
+        # inputs: (batch * length, num_hidden)
+        inputs = mx.sym.reshape(data=inputs, shape=(-3, -1))
+
+        # project with 1 large matrix
+        # combined: (batch * length, depth * 3)
+        combined = mx.sym.FullyConnected(data=inputs,
+                                         weight=self.w_i2h,
+                                         bias=self.b_i2h,
+                                         num_hidden=self.depth_att * 3,
+                                         name="%sqkv_transform" % self.prefix)
+
+        # split batch and time dimension
+        # combined: (batch, length, depth * 3)
+        combined = mx.sym.reshape(data=combined, shape=(-1, length, self.depth_att * 3))
+
+        # split into query, keys and values
+        # q/k/v: (batch, length, depth)
+        # NOTE: requires depth to be equal across all 3 parts.
+        q, k, v = mx.sym.split(data=combined, num_outputs=3, axis=2)
+
+        # q/k/v: (batch * heads, length, depth/heads)
+        q = split_heads(q, length, self.heads)
+        k = split_heads(k, length, self.heads)
+        v = split_heads(v, length, self.heads)
+        # scale by sqrt(depth_per_head)
+        q *= self.depth_per_head ** -0.5
+
+        # (batch*heads, length, depth_per_head) X (batch*heads, depth_per_head, length)
+        #   -> (batch*heads, length, length)
+        logits = mx.sym.batch_dot(lhs=q, rhs=k, transpose_b=True)
+
+        # mask variable sequence-length (unfortunately requires time-major data)
+        # logits: (length, batch * heads, length)
+        logits = mx.sym.swapaxes(data=logits, dim1=0, dim2=1)
+        logits = mx.sym.SequenceMask(data=logits,
+                                     use_sequence_length=True,
+                                     sequence_length=broadcast_lengths(inputs_length, self.heads),
+                                     value=-99999999.)
+        # logits: (batch * heads, length, length)
+        logits = mx.sym.swapaxes(data=logits, dim1=0, dim2=1)
+
+        # weights: (batch * heads, length, length)
+        weights = mx.sym.softmax(logits)
+
+        if self.dropout > 0.0:
+            weights = mx.sym.Dropout(weights, p=self.dropout)
+
+        if self.use_loop:
```

Comment:
no, the for loop was just my first attempt. I will leave it in a little bit more but eventually we can remove it